feat(HMS-3686): Update position of blueprint buttons

avitova opened the pull request:
This is a followup of #1631 and an update according to our new mockups. :)

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -171,16 +207,53 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
       {itemCount === 0 && <EmptyImagesTable />}
       {itemCount > 0 && (
         <>
+          {/*@ts-ignore*/}
```
Leftover I think, let's remove.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -93,6 +105,30 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+  const onBuildHandler = async () => {
+    selectedBlueprint && (await buildBlueprint({ id: selectedBlueprint }));
+  };
+  const [isOpen, setIsOpen] = useState<boolean>(false);
```
We don't need to specify `<boolean>` here, Typescript can infer it since the hook was initialized with `false`.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -93,6 +105,30 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+    setIsOpen(!isOpen);
+  };
+
+  const [showDeleteModal, setShowDeleteModal] = React.useState(false);
```
`useState` has been imported, so you don't need to use `React.`.

lucasgarfield reviewed (changes requested):
This is looking great! One oversight, the changes leaked out of experimental. Here's what I see running `npm run stage-beta`:

![image](https://github.com/osbuild/image-builder-frontend/assets/95542540/308d0270-a7c6-45e2-815e-030bfac892c1)

You'll just need to put all of this behind the experimental flag. You can use conditional rendering like so: `{experimentalFlag && ... }`... you should be able to find plenty of examples of this in the code already.

avitova commented:
Thank you for your review @lucasgarfield 
I have rebased, addressed the comments, fixed a small bug (when deleting a blueprint, we should set the selectedBlueprint to undefined), and put back the Create image button since that is needed out of experimental. :)

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -84,6 +100,31 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
     setPage(1);
     setPerPage(perPage);
   };
+  const [buildBlueprint, { isLoading: imageBuildLoading }] =
```
We can move this into a new `<BlueprintActionsMenu>` component.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -84,6 +100,31 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+  const [buildBlueprint, { isLoading: imageBuildLoading }] =
+    useComposeBlueprintMutation();
+
+  const onBuildHandler = async () => {
```
We can move this into a new `<BuildImagesButton>` component.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -84,6 +100,31 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+  };
+
+  const [showDeleteModal, setShowDeleteModal] = useState(false);
+  const [deleteBlueprint] = useDeleteBlueprintMutation({
```
We can move this to `<DeleteBlueprintModal>` I think.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -84,6 +100,31 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+  const [deleteBlueprint] = useDeleteBlueprintMutation({
+    fixedCacheKey: 'delete-blueprint',
+  });
+  const handleDelete = async () => {
```
We can move this to `<DeleteBlueprintModal>` I think.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -84,6 +100,31 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
+      setSelectedBlueprint(undefined);
+    }
+  };
+  const onDeleteClose = () => {
```
We can move this to `<DeleteBlueprintModal>` I think.

lucasgarfield reviewed (changes requested):
Thanks for the updates! I have some opinionated nitpicks next. I’m trying not to block on things that are too nitpicky but I think these won’t be too hard and should be worth it.

I would like to extract some of the JSX into new React components, and move some code (handlers and RTKQ hooks) deeper where it is closer to where it’s needed. This will help keep the scope of `<ImagesTable>` smaller, and keeping the handlers closer to where they called makes the code easier to read and understand. In general I try to avoid passing down props as much as possible, and since RTKQ hooks are cached I prefer to just call them in the component they’re needed in. So with that in mind, I hope you think these requests are reasonable. :sweat_smile:

avitova commented:
Thank you, Lucas, for the amount of suggestions. I think it all makes sense, and the code looks cleaner. All of the mentioned components are in a sense logical units, so I like the idea of not keeping this all in the images table. :) I decomposed based on your suggestion.

lucasgarfield commented on src/Components/ImagesTable/ImagesTable.tsx:
```diff
@@ -186,6 +215,7 @@ const ImagesTable = ({ selectedBlueprint }: ImageTableProps) => {
                 isCompact
               />
             </ToolbarItem>
+            )
```
This `)` is getting rendered to the right of the pagination :wink: 

![image](https://github.com/osbuild/image-builder-frontend/assets/95542540/08ff0743-1e4e-4691-99ed-1360e32989d2)

lucasgarfield commented:
Thanks for refactoring, @avitova! This looks really nice! 

Please just fix the tiny bug with the `)` I mentioned and I'll approve and merge.

We will probably also need to do a follow-up and convert these components to use the Redux store when https://github.com/osbuild/image-builder-frontend/pull/1712 gets merged.

lucasgarfield reviewed (approved):
Thank you!
